rpisense-core.c

```diff
@@ -60,56 +47,75 @@ static int rpisense_probe(struct i2c_client *i2c,
 	rpisense->dev = &i2c->dev;
 	rpisense->i2c_client = i2c;
 
-	ret = rpisense_reg_read(rpisense, RPISENSE_WAI);
-	if (ret > 0) {
-		if (ret != RPISENSE_ID)
-			return -EINVAL;
-	} else {
+
+	ret = i2c_smbus_read_byte_data(rpisense->i2c_client, RPISENSE_WAI);
+	if (ret < 0)
 		return ret;
-	}
-	ret = rpisense_reg_read(rpisense, RPISENSE_VER);
+
+	if (ret != RPISENSE_ID)
+			return -EINVAL;
+
+	ret = i2c_smbus_read_byte_data(rpisense->i2c_client, RPISENSE_VER);
 	if (ret < 0)
 		return ret;
 
 	dev_info(rpisense->dev,
 		 "Raspberry Pi Sense HAT firmware version %i\n", ret);
 
-	rpisense_js = &rpisense->joystick;
-	rpisense_js->keys_desc = devm_gpiod_get(&i2c->dev,
-						"keys-int", GPIOD_IN);
-	if (IS_ERR(rpisense_js->keys_desc)) {
-		dev_warn(&i2c->dev, "Failed to get keys-int descriptor.\n");
-		return PTR_ERR(rpisense_js->keys_desc);
+	rpisense->joystick.pdev = rpisense_client_dev_register(rpisense,
+							       "rpi-sense-js");
+
+	if(IS_ERR(rpisense->joystick.pdev)) {
+		dev_err(rpisense->dev, "failed to register rpisense-js");
+		return PTR_ERR(rpisense->joystick.pdev);
+	}
+
+	rpisense->framebuffer.pdev = rpisense_client_dev_register(rpisense,
+								  "rpi-sense-fb");
+
+	if(IS_ERR(rpisense->framebuffer.pdev)) {
+		dev_err(rpisense->dev, "failed to register rpisense-fb");
+		return PTR_ERR(rpisense->framebuffer.pdev);
 	}
-	rpisense_client_dev_register(rpisense, "rpi-sense-js",
-				     &(rpisense->joystick.pdev));
-	rpisense_client_dev_register(rpisense, "rpi-sense-fb",
-				     &(rpisense->framebuffer.pdev));
 
 	return 0;
 }
 
-static int rpisense_remove(struct i2c_client *i2c)
+static struct platform_device *
+rpisense_client_dev_register(struct rpisense *rpisense, const char *name)
 {
-	struct rpisense *rpisense = i2c_get_clientdata(i2c);
+	long ret = -ENOMEM;
+	struct platform_device *pdev = platform_device_alloc(name, -1);
+	if(pdev == NULL)
+		goto alloc_fail;
 
-	platform_device_unregister(rpisense->joystick.pdev);
-	return 0;
-}
+	pdev->dev.parent = rpisense->dev;
+	platform_set_drvdata(pdev, rpisense);
 
+	ret = platform_device_add(pdev);
+	if (ret != 0)
+		goto add_fail;
 
-s32 rpisense_reg_read(struct rpisense *rpisense, int reg)
+	ret = devm_add_action_or_reset(rpisense->dev,
+		(void *)platform_device_unregister, pdev);
+	if(ret != 0)
+		goto alloc_fail;
+
+	return pdev;
+
+add_fail:
+	platform_device_put(pdev);
+alloc_fail:
+	return ERR_PTR(ret);
+}
+
+int rpisense_get_joystick_state(struct rpisense *rpisense)
```

Comment:
do you think this might better belong in rpisense-js?

Reply:
see reasoning above for why I moved the constants into core